mllib/src/main/scala/org/apache/spark/mllib/clustering/DpMeans.scala

```diff
@@ -0,0 +1,279 @@
+  def getMaxClusterCount: Int = maxClusterCount
+
+  /**
+   * Perform DP-means clustering
+   */
+  def run(data: RDD[Vector]): DpMeansModel = {
+    if (data.getStorageLevel == StorageLevel.NONE) {
+      logWarning("The input data is not directly cached, which may hurt performance if its"
+        + " parent RDDs are also uncached.")
+    }
+
+    // Compute squared norms and cache them.
+    val norms = data.map(Vectors.norm(_, 2.0))
+    norms.persist()
+    val zippedData = data.zip(norms).map {
+      case (v, norm) => new VectorWithNorm(v, norm)
+    }
+
+    // Implementation of DP-means algorithm.
+    var iteration = 0
+    var previousK = 1
+    var covered = false
+    var converged = false
+    var localCenters = Array.empty[VectorWithNorm]
+    val globalCenters = ArrayBuffer.empty[VectorWithNorm]
+
+    // Execute clustering until the maximum number of iterations is reached
+    // or the cluster centers have converged.
+    while (iteration < maxIterations && !converged) {
+
+      // This loop simulates the cluster assignment step by iterating through the data
+      // points until no new clusters are generated. A new cluster is formed whenever
+      // the distance from a data point to all the existing cluster centroids is greater
+      // than lambdaValue. The cluster centers returned by generateNewCenters(), called
+      // localCenters, are sent to the master. The cluster creation process is performed
+      // locally on these localCenters and the resulting cluster centers are added to
+      // the existing set of global centers.
+      while (!covered) {
+        val findLocalCenters = zippedData
+          .mapPartitions(h =>
+              DpMeans.generateNewCenters(h, globalCenters, lambdaValue)
+          ).persist()
+        val localCentersCount = findLocalCenters.count()
+        logInfo(" Local centers count :: " + localCentersCount)
+        if (localCentersCount > getMaxClusterCount) {
+          logInfo(" Local centers count " + localCentersCount + " exceeds user expectation")
+          sys.exit()
+        }
+        localCenters = findLocalCenters.collect()
+        if (localCenters.isEmpty) {
+          covered = true
+        }
+        // Promote a local cluster center to a global center
+        else {
+          val newGlobalCenters = DpMeans.generateNewCenters(localCenters.iterator,
+            ArrayBuffer.empty[VectorWithNorm], lambdaValue)
+          globalCenters ++= newGlobalCenters
+          logInfo(" Number of global centers :: " + globalCenters.length)
+        }
+      }
+
+      // Find the sum and count of points belonging to each cluster
+      case class WeightedPoint(vector: Vector, count: Long)
+      val mergeClusters = (x: WeightedPoint, y: WeightedPoint) => {
+        axpy(1.0, y.vector, x.vector)
+        WeightedPoint(x.vector, x.count + y.count)
+      }
+
+      val dims = globalCenters(0).vector.size
+      val clusterStat = zippedData.mapPartitions { points =>
```

Comment:
@FlytxtRnD sure! Tthis would work like this. However, I haven't onfirm carefully yet.
https://github.com/yu-iskw/spark/commit/4070ae666eade692a129a95f5635047b39ccbed7